internal/exec/config_fetcher.go

```diff
@@ -0,0 +1,146 @@
+		newCfg, err := f.fetchReferencedConfig(cfgRef)
+		if err != nil {
+			return types.Config{}, err
+		}
+
+		// Replace the HTTP client in the fetcher to be configured with the
+		// timeouts of the new config
+		err = f.Fetcher.UpdateHttpTimeoutsAndCAs(newCfg.Ignition.Timeouts, newCfg.Ignition.Security.TLS.CertificateAuthorities, newCfg.Ignition.Proxy)
+		if err != nil {
+			return types.Config{}, err
+		}
+
+		return f.RenderConfig(newCfg)
+	}
+
+	appendedCfg := cfg
+	for _, cfgRef := range cfg.Ignition.Config.Merge {
+		newCfg, err := f.fetchReferencedConfig(cfgRef)
+		if err != nil {
+			return types.Config{}, err
+		}
+
+		// Merge the old config with the new config before the new config has
+		// been rendered, so we can use the new config's timeouts and CAs when
+		// fetching more configs.
+		cfgForFetcherSettings := latest.Merge(appendedCfg, newCfg)
+		err = f.Fetcher.UpdateHttpTimeoutsAndCAs(cfgForFetcherSettings.Ignition.Timeouts, cfgForFetcherSettings.Ignition.Security.TLS.CertificateAuthorities, cfgForFetcherSettings.Ignition.Proxy)
+		if err != nil {
+			return types.Config{}, err
+		}
+
+		newCfg, err = f.RenderConfig(newCfg)
+		if err != nil {
+			return types.Config{}, err
+		}
+
+		appendedCfg = latest.Merge(appendedCfg, newCfg)
+	}
+	return appendedCfg, nil
+}
+
+// fetchReferencedConfig fetches and parses the requested config.
+// cfgRef.Source must not be nil
+func (f *ConfigFetcher) fetchReferencedConfig(cfgRef types.Resource) (types.Config, error) {
+	// this is also already checked at validation time
+	if cfgRef.Source == nil {
+		f.Logger.Crit("invalid referenced config: %v", errors.ErrSourceRequired)
+		return types.Config{}, errors.ErrSourceRequired
+	}
+	u, err := url.Parse(*cfgRef.Source)
+	if err != nil {
+		return types.Config{}, err
+	}
+	var headers http.Header
+	if cfgRef.HTTPHeaders != nil && len(cfgRef.HTTPHeaders) > 0 {
+		headers, err = cfgRef.HTTPHeaders.Parse()
+		if err != nil {
+			return types.Config{}, err
+		}
+	}
+	compression := ""
+	if cfgRef.Compression != nil {
+		compression = *cfgRef.Compression
+	}
+	rawCfg, err := f.Fetcher.FetchToBuffer(*u, resource.FetchOptions{
+		Headers:     headers,
+		Compression: compression,
+	})
+	if err != nil {
+		return types.Config{}, err
+	}
+
+	hash := sha512.Sum512(rawCfg)
+	if u.Scheme != "data" {
+		f.Logger.Debug("fetched referenced config at %s with SHA512: %s", *cfgRef.Source, hex.EncodeToString(hash[:]))
+	} else {
+		// data url's might contain secrets
+		f.Logger.Debug("fetched referenced config from data url with SHA512: %s", hex.EncodeToString(hash[:]))
+	}
+
+	if err := util.AssertValid(cfgRef.Verification, rawCfg); err != nil {
+		return types.Config{}, err
+	}
+
+	cfg, r, err := config.Parse(rawCfg)
+	f.Logger.LogReport(r)
+	if err != nil {
+		return types.Config{}, err
+	}
+
+	f.State.FetchedConfigs = append(f.State.FetchedConfigs, state.FetchedConfig{
+		Kind:       "user",
+		Source:     u.Path,
+		Referenced: true,
+	})
```

Comment:
This block moved within the function.  It's a valid bugfix, but let's either do that in a separate commit or at least call it out in the commit message.